scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala

```diff
@@ -334,11 +334,218 @@ object OptimizationRules {
       Binary(recurse(hj.left), rightLit,
         { (ltp: TypedPipe[(K, V)], rtp: TypedPipe[(K, V2)]) =>
           rtp match {
-            case ReduceStepPipe(hg: HashJoinable[K, V2]) =>
+            case ReduceStepPipe(hg: HashJoinable[K @unchecked, V2 @unchecked]) =>
               HashCoGroup(ltp, hg, joiner)
             case otherwise =>
               HashCoGroup(ltp, IdentityReduce(ordK, otherwise, None, Nil), joiner)
           }
         })
     }
+
+
+  /////////////////////////////
+  //
+  // Here are some actual rules for simplifying TypedPipes
+  //
+  /////////////////////////////
+
+  object ComposeFlatMap extends PartialRule[TypedPipe] {
```

Comment:
wondering if it would be a good idea to include a short snippet of code as a comment on the top of each rule showing which scenario it gets applied in?